### DISCUSSION TITLE
testing recursive async operation with setTimeout

### DISCUSSION THREAD

**Discussion by ssutar** [vitest-dev/vitest#5178] [Q&A]
### Describe the bug

I was writing the test for a flow wherein I need to poll an API until it returns `success` or `failure` repeatedly after some time interval. Though the code works as expected in the browser, I am not able to figure out the tests. It times out even though using `faketimers` I also tried most of the combinations of runAllTimers and flushPromises however nothing seem to work

See the `not work` test in the stackblitz link below:

### Reproduction

https://stackblitz.com/edit/vitest-dev-vitest-gketdn?file=test%2Frecursive.test.ts

### System Info

```shell
System:
    OS: Linux 5.0 undefined
    CPU: (8) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Memory: 0 Bytes / 0 Bytes
    Shell: 1.0 - /bin/jsh
  Binaries:
    Node: 18.18.0 - /usr/local/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 10.2.3 - /usr/local/bin/npm
    pnpm: 8.14.0 - /usr/local/bin/pnpm
  npmPackages:
    @vitest/ui: latest => 1.2.2 
    vite: latest => 5.1.0 
    vitest: latest => 1.2.2
```

### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by ssutar** [comment]
It works when `setTimeout` is mocked

```
vi.spyOn(globalThis, 'setTimeout').mockImplementation((fn: () => any) => {
      return fn();
    });
```

**Comment by hi-ogawa** [comment]
It's likely that there is an issue in your code (or your expectation about how promise, async/await, etc... works in JS).

I cannot tell you how to debug your code exactly (unless reproduction is further minimized), but the issue is that your code is stuck at `await`, so your code is not reaching `vi.advanceTimersByTime` nor `vi.runAllTimers()` 

```tsx
    const result = await callRecursivePromise(1);
    vi.advanceTimersByTime(POLL_TIMEOUT + 100);
    vi.runAllTimers();
```

As a simplified example, the following test doesn't work for the same reason. Hopefully this will give you some idea:

https://stackblitz.com/edit/vitest-dev-vitest-4h8ext?file=src%2Frepro.test.ts

```ts
import { it, vi, beforeEach } from 'vitest';

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

beforeEach(() => {
  vi.useFakeTimers();
});

it('stuck', async () => {
  await sleep(1000); // it's sleeping forever 
  vi.runAllTimers();
});
```
